toys/LisExpr.chpl

```diff
@@ -261,32 +325,50 @@ module LisExpr
       tokenize the prog
     */
     proc tokenize(line: string) {
-        return line.replace("("," ( ").replace(")"," ) ").split();
+        // Want:
+        //   var l: list(string) = line.replace("("," ( ").replace(")"," ) ").split()
+        // Workaround (see https://github.com/chapel-lang/chapel/issues/16166):
+
+        var l: list(string);
+        for token in line.replace("("," ( ").replace(")"," ) ").split() do
+          l.append(token);
+        return l;
     }
     
     /*
       parse, check, and validate code and all symbols in the tokenized prog
     */ 
     proc parse(line: string): owned GenListValue throws {
-        return read_from_tokens(tokenize(line));
+        // Want:
+        //   return read_from_tokens(tokenize(line));
+        //
+        // Workaround (see https://github.com/chapel-lang/chapel/issues/16170):
+
+        var l: list(string) = tokenize(line);
+        return read_from_tokens(l);
     }
     
     /*
       parse throught the list of tokens generating the parse tree / AST
       as a list of atoms and lists
     */
-    proc read_from_tokens(tokens: [?D] string): owned GenListValue throws {
+    proc read_from_tokens(ref tokens: list(string)): owned GenListValue throws {
         if (tokens.size == 0) then
             throw new owned ErrorWithMsg("SyntaxError: unexpected EOF");
-        var token = tokens.pop_front();
+
+        // Open Q: If we were to parse from the back of the string to the
+        // front, could this be more efficient since popping from the
+        // front of a list is an expensive operation?
+
+        var token = tokens.pop(0);
         if (token == "(") {
             var L: GenList;
-            while (tokens[D.low] != ")") {
+            while (tokens.first() != ")") {
                 L.append(read_from_tokens(tokens));
                 if (tokens.size == 0) then
                     throw new owned ErrorWithMsg("SyntaxError: unexpected EOF");
             }
-            tokens.pop_front(); // pop off ")"
+            tokens.pop(0); // pop off ")"
```

Comment:
With all of Chapel's index-neutral features, it seems a shame that the only way to remove an element from the front of a list is by using an explicit index value...  This might be worth another issue as well.

Reply:
We could totally add a method called `popFront` to list.

Reply:
`popFirst` ?

Reply:
does `pop` equal `popLast`?

Reply:
Yeah, at present.  I personally like popFirst and popLast better than popHead and popTail because they seem less ambiguous w.r.t. order (and, depending on the implementation, expense).  I believe lists support .first and .last queries already, so this would also be more uniform.

Reply:
so is `last` the first thing you appended? and `first` the last thing you appended? or the other way around?

Reply:
For me, the key is to think of lists as being neither stack nor queue.  Do I know what a list is, in real life?  (Yeah).  Do I know what the first thing is?  (Yeah).  Do I know what the last thing is?  (Yeah).  What does appending something to the list do?  (Add a new entry to the end / after the last thing).  Pop is admittedly weird given all this and wouldn't have been my choice for a list operation.  I don't recall what the rationale there was, though I expect it related to following in some other language's footsteps and/or "What is the cheapest thing to pop off of a list [for our current implementation]".  That said, we're discussing having multiple implementation options under the hood, which I think would make simple `pop()` go away and require `popFirst` vs. `popLast` vs. `pop(i)`.

Reply:
I captured some of my takeaways from this conversation here:  https://github.com/chapel-lang/chapel/issues/16194